
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>baseqdrops.tagging.prime3 &#8212; baseqDrops 0.01 documentation</title>
    <link rel="stylesheet" href="../../../_static/pyramid.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/customize.css" type="text/css" />
    <script type="text/javascript" src="../../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Neuton&amp;subset=latin" type="text/css" media="screen" charset="utf-8" />
<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Nobile:regular,italic,bold,bolditalic&amp;subset=latin" type="text/css" media="screen" charset="utf-8" />
<!--[if lte IE 6]>
<link rel="stylesheet" href="../../../_static/ie6.css" type="text/css" media="screen" charset="utf-8" />
<![endif]-->

  </head><body>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">baseqDrops 0.01 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" accesskey="U">Module code</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for baseqdrops.tagging.prime3</h1><div class="highlight"><pre>
<span></span><span class="c1"># -*- coding: utf-8 -*-</span>
<span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">bisect</span><span class="o">,</span> <span class="nn">sys</span><span class="o">,</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">.bam_parser</span> <span class="k">import</span> <span class="n">get_cigar_segments</span>
<span class="kn">from</span> <span class="nn">baseq.mgt.config</span> <span class="k">import</span> <span class="n">get_config</span>

<div class="viewcode-block" id="tagging_reads"><a class="viewcode-back" href="../../../函数介绍.html#baseqdrops.tagging_reads">[docs]</a><span class="k">def</span> <span class="nf">tagging_reads</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">bam</span><span class="p">,</span> <span class="n">outpath</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    **tagging_reads** will read the bam file which sorted by header or barcode, report the UMIs and reads counts for</span>
<span class="sd">    each expressed genes of a barcode.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        genome: The name of genome (hg38);</span>
<span class="sd">        bam: The bam file (should sorted by header);</span>
<span class="sd">        outpath: The path of genes expression files;</span>
<span class="sd">    </span>
<span class="sd">    Todo:</span>
<span class="sd">        * Check the existence of reference files;</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">        path: A file for tagged reads;</span>
<span class="sd">        ::</span>
<span class="sd">            BARCODE#{}</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">#Remove the Already Exists File</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">outpath</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">outpath</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;[info] File write to </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">outpath</span><span class="p">))</span>

    <span class="n">cellranger_refs</span> <span class="o">=</span> <span class="n">get_config</span><span class="p">(</span><span class="s2">&quot;Drops&quot;</span><span class="p">,</span> <span class="s2">&quot;cellranger_ref_&quot;</span><span class="o">+</span><span class="n">genome</span><span class="p">)</span>
    <span class="n">star_index_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cellranger_refs</span><span class="p">,</span> <span class="s2">&quot;star&quot;</span><span class="p">)</span>

    <span class="c1">#read star index</span>
    <span class="n">chrom_name_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">star_index_dir</span><span class="p">,</span> <span class="s2">&quot;chrName.txt&quot;</span><span class="p">)</span>
    <span class="n">chrom_start_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">star_index_dir</span><span class="p">,</span> <span class="s2">&quot;chrStart.txt&quot;</span><span class="p">)</span>
    <span class="n">tx_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">star_index_dir</span><span class="p">,</span> <span class="s2">&quot;transcriptInfo.tab&quot;</span><span class="p">)</span>
    <span class="n">ex_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">star_index_dir</span><span class="p">,</span> <span class="s2">&quot;exonInfo.tab&quot;</span><span class="p">)</span>

    <span class="c1">#gene transc</span>
    <span class="n">gene_transc</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cellranger_refs</span><span class="p">,</span> <span class="s2">&quot;genes_transcript.txt&quot;</span><span class="p">)</span>

    <span class="n">tx_dtype</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;tx_ids&#39;</span><span class="p">,</span> <span class="s1">&#39;tx_starts&#39;</span><span class="p">,</span> <span class="s1">&#39;tx_ends&#39;</span><span class="p">,</span> <span class="s1">&#39;tx_max_ends&#39;</span><span class="p">,</span> <span class="s1">&#39;tx_strands&#39;</span><span class="p">,</span>
                          <span class="s1">&#39;tx_num_exons&#39;</span><span class="p">,</span> <span class="s1">&#39;tx_break_idxs&#39;</span><span class="p">),</span>
                <span class="s1">&#39;formats&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;object&#39;</span><span class="p">,</span> <span class="s1">&#39;u8&#39;</span><span class="p">,</span> <span class="s1">&#39;u8&#39;</span><span class="p">,</span> <span class="s1">&#39;u8&#39;</span><span class="p">,</span> <span class="s1">&#39;u1&#39;</span><span class="p">,</span> <span class="s1">&#39;u2&#39;</span><span class="p">,</span> <span class="s1">&#39;u4&#39;</span><span class="p">)}</span>
    <span class="n">ex_dtype</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;ex_starts&#39;</span><span class="p">,</span> <span class="s1">&#39;ex_ends&#39;</span><span class="p">,</span> <span class="s1">&#39;ex_cum_lengths&#39;</span><span class="p">),</span>
        <span class="s1">&#39;formats&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;u8&#39;</span><span class="p">,</span> <span class="s1">&#39;u8&#39;</span><span class="p">,</span> <span class="s1">&#39;u8&#39;</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="n">gene_tx_dtype</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span> <span class="s1">&#39;tx&#39;</span><span class="p">,</span> <span class="s1">&#39;genename&#39;</span><span class="p">),</span>
        <span class="s1">&#39;formats&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;object&#39;</span><span class="p">,</span> <span class="s1">&#39;object&#39;</span><span class="p">,</span> <span class="s1">&#39;object&#39;</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="n">chrom_names</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="n">chrom_name_path</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;object&#39;</span><span class="p">)</span>
    <span class="n">chrom_starts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="n">chrom_start_path</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;u8&#39;</span><span class="p">)</span>
    <span class="n">tx_info</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="n">tx_path</span><span class="p">,</span> <span class="n">tx_dtype</span><span class="p">,</span> <span class="n">skiprows</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">unpack</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">ex_info</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="n">ex_path</span><span class="p">,</span> <span class="n">ex_dtype</span><span class="p">,</span> <span class="n">skiprows</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">unpack</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">ex_tx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="n">gene_transc</span><span class="p">,</span> <span class="n">gene_tx_dtype</span><span class="p">,</span> <span class="n">unpack</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">tx_starts</span> <span class="o">=</span> <span class="n">tx_info</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">chrom_bins</span> <span class="o">=</span> <span class="p">[</span><span class="n">bisect</span><span class="o">.</span><span class="n">bisect_left</span><span class="p">(</span><span class="n">tx_starts</span><span class="p">,</span> <span class="n">cs</span><span class="p">)</span> <span class="k">for</span> <span class="n">cs</span> <span class="ow">in</span> <span class="n">chrom_starts</span><span class="p">]</span>
    <span class="n">chrom_info</span> <span class="o">=</span> <span class="n">chrom_names</span><span class="p">,</span> <span class="n">chrom_starts</span><span class="p">,</span> <span class="n">chrom_bins</span>

    <span class="n">ex_starts</span><span class="p">,</span> <span class="n">ex_ends</span><span class="p">,</span> <span class="n">ex_cum_lengths</span> <span class="o">=</span> <span class="n">ex_info</span>
    <span class="n">ex_breaks</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="mi">2</span> <span class="o">*</span> <span class="n">ex_starts</span><span class="o">.</span><span class="n">size</span><span class="p">,),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">ex_starts</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span>
    <span class="n">ex_breaks</span><span class="p">[</span><span class="mi">0</span><span class="p">::</span><span class="mi">2</span><span class="p">]</span> <span class="o">=</span> <span class="n">ex_starts</span>
    <span class="n">ex_breaks</span><span class="p">[</span><span class="mi">1</span><span class="p">::</span><span class="mi">2</span><span class="p">]</span> <span class="o">=</span> <span class="n">ex_ends</span>

    <span class="n">tx2gene</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ex_tx</span><span class="p">[</span><span class="mi">0</span><span class="p">])):</span>
        <span class="n">tx2gene</span><span class="p">[</span><span class="n">ex_tx</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="n">idx</span><span class="p">]]</span> <span class="o">=</span> <span class="n">ex_tx</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="n">idx</span><span class="p">]</span>

    <span class="n">transcript_ids</span><span class="p">,</span> <span class="n">transcript_starts</span><span class="p">,</span> <span class="n">transcript_ends</span><span class="p">,</span> <span class="n">transcript_max_ends</span><span class="p">,</span> <span class="n">transcript_strands</span><span class="p">,</span> <span class="n">transcript_num_exons</span><span class="p">,</span> <span class="n">transcript_exon_break_idx</span> <span class="o">=</span> <span class="n">tx_info</span>

    <span class="c1">#This function returns the information: the genename, exon or intron, antisense?</span>
    <span class="c1">#It should return a single transcript and Gene</span>

    <span class="k">def</span> <span class="nf">align_to_transcriptome</span><span class="p">(</span><span class="n">read_tid</span><span class="p">,</span> <span class="n">read_pos</span><span class="p">,</span> <span class="n">read_alen</span><span class="p">):</span>
        <span class="n">ref_offset</span> <span class="o">=</span> <span class="n">chrom_starts</span><span class="p">[</span><span class="n">read_tid</span><span class="p">]</span>
        <span class="n">clipped_read_start</span> <span class="o">=</span> <span class="n">ref_offset</span> <span class="o">+</span> <span class="n">read_pos</span>
        <span class="n">clipped_read_end</span> <span class="o">=</span> <span class="n">clipped_read_start</span> <span class="o">+</span> <span class="n">read_alen</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="n">tx</span> <span class="o">=</span> <span class="n">bisect</span><span class="o">.</span><span class="n">bisect</span><span class="p">(</span><span class="n">transcript_starts</span><span class="p">,</span> <span class="n">clipped_read_end</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>

        <span class="c1"># read is at the extreme start / end of transcriptome</span>
        <span class="k">if</span> <span class="n">tx</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span> <span class="ow">or</span> <span class="n">tx</span> <span class="o">==</span> <span class="n">transcript_starts</span><span class="o">.</span><span class="n">size</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">{}</span>
        <span class="n">tx_hits</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">while</span> <span class="n">tx</span> <span class="o">&gt;=</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">clipped_read_start</span> <span class="o">&lt;=</span> <span class="nb">max</span><span class="p">(</span><span class="n">transcript_max_ends</span><span class="p">[</span><span class="n">tx</span><span class="p">],</span> <span class="n">transcript_ends</span><span class="p">[</span><span class="n">tx</span><span class="p">]):</span>
            <span class="k">if</span> <span class="n">clipped_read_start</span> <span class="o">&lt;=</span> <span class="n">transcript_ends</span><span class="p">[</span><span class="n">tx</span><span class="p">]:</span>
                <span class="n">tx_hits</span><span class="p">[</span><span class="n">transcript_ids</span><span class="p">[</span><span class="n">tx</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="n">clipped_read_start</span><span class="o">-</span><span class="n">transcript_starts</span><span class="p">[</span><span class="n">tx</span><span class="p">],</span>
                    <span class="s1">&#39;exons&#39;</span><span class="p">:</span> <span class="n">transcript_num_exons</span><span class="p">[</span><span class="n">tx</span><span class="p">],</span>
                    <span class="s1">&#39;exonIdx&#39;</span><span class="p">:</span> <span class="n">transcript_exon_break_idx</span><span class="p">[</span><span class="n">tx</span><span class="p">],</span>
                    <span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="n">tx2gene</span><span class="p">[</span><span class="n">transcript_ids</span><span class="p">[</span><span class="n">tx</span><span class="p">]],</span>
                    <span class="s1">&#39;strand&#39;</span><span class="p">:</span> <span class="n">transcript_strands</span><span class="p">[</span><span class="n">tx</span><span class="p">]</span>
                <span class="p">}</span>
            <span class="n">tx</span> <span class="o">-=</span> <span class="mi">1</span>
        <span class="k">return</span> <span class="n">tx_hits</span>


    <span class="k">def</span> <span class="nf">align2exon</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">ex_idx</span><span class="p">,</span> <span class="n">ex_idx2</span><span class="p">):</span>
        <span class="n">ex_start</span> <span class="o">=</span> <span class="n">bisect</span><span class="o">.</span><span class="n">bisect</span><span class="p">(</span><span class="n">ex_breaks</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">lo</span><span class="o">=</span><span class="mi">2</span> <span class="o">*</span> <span class="n">ex_idx</span><span class="p">,</span> <span class="n">hi</span><span class="o">=</span><span class="mi">2</span> <span class="o">*</span> <span class="p">(</span><span class="n">ex_idx2</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="n">ex_end</span> <span class="o">=</span> <span class="n">bisect</span><span class="o">.</span><span class="n">bisect_left</span><span class="p">(</span><span class="n">ex_breaks</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">lo</span><span class="o">=</span><span class="nb">max</span><span class="p">(</span><span class="n">ex_start</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="n">hi</span><span class="o">=</span><span class="mi">2</span> <span class="o">*</span> <span class="p">(</span><span class="n">ex_idx2</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="n">exon_len</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">dist_to_end</span> <span class="o">=</span> <span class="mi">10000</span>

        <span class="k">if</span> <span class="n">ex_start</span><span class="o">%</span><span class="mi">2</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
            <span class="n">exon_len</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">ex_breaks</span><span class="p">[</span><span class="n">ex_start</span><span class="o">+</span><span class="mi">1</span><span class="p">],</span> <span class="n">end</span><span class="p">)</span> <span class="o">-</span> <span class="nb">max</span><span class="p">(</span><span class="n">ex_breaks</span><span class="p">[</span><span class="n">ex_start</span><span class="p">],</span> <span class="n">start</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">ex_end</span><span class="o">%</span><span class="mi">2</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
            <span class="n">exon_len</span> <span class="o">=</span> <span class="n">end</span> <span class="o">-</span> <span class="n">ex_breaks</span><span class="p">[</span><span class="n">ex_start</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span>

        <span class="c1">##Calculate The distance of ex_Start to the End of Genes...</span>
        <span class="c1">## We should Seprate by the Strand Of the Genes...</span>

        <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">exon_to_geneend</span> <span class="o">=</span> <span class="n">ex_breaks</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">ex_start</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span><span class="o">*</span><span class="mi">2</span> <span class="o">+</span><span class="mi">2</span><span class="p">:</span> <span class="mi">2</span> <span class="o">*</span> <span class="p">(</span><span class="n">ex_idx2</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">exon_to_geneend</span> <span class="o">=</span> <span class="n">ex_breaks</span><span class="p">[</span><span class="mi">2</span> <span class="o">*</span> <span class="n">ex_idx</span> <span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">ex_start</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span> <span class="o">*</span> <span class="mi">2</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">exon_to_geneend</span><span class="p">)</span><span class="o">%</span><span class="mi">2</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
            <span class="n">dist_to_end</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">exon_to_geneend</span><span class="p">[</span><span class="mi">1</span><span class="p">::</span><span class="mi">2</span><span class="p">])</span> <span class="o">-</span> <span class="nb">sum</span><span class="p">(</span><span class="n">exon_to_geneend</span><span class="p">[::</span><span class="mi">2</span><span class="p">])</span>

        <span class="k">return</span> <span class="n">ex_start</span><span class="p">,</span> <span class="n">ex_end</span><span class="p">,</span> <span class="n">exon_len</span><span class="p">,</span> <span class="n">dist_to_end</span>

    <span class="c1">##Read The Bamfile</span>
    <span class="c1">##Iterate the bam file</span>

    <span class="kn">import</span> <span class="nn">pysam</span><span class="o">,</span> <span class="nn">itertools</span>
    <span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">Counter</span>

    <span class="n">bam_file</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">Samfile</span><span class="p">(</span><span class="n">bam</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
    <span class="n">genome_bam_iter</span> <span class="o">=</span> <span class="n">itertools</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">bam_file</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">read</span><span class="p">:</span> <span class="n">read</span><span class="o">.</span><span class="n">qname</span><span class="p">)</span>

    <span class="n">res</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">previous_barcode</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

    <span class="k">for</span> <span class="n">qname</span><span class="p">,</span> <span class="n">reads_iter</span> <span class="ow">in</span> <span class="n">genome_bam_iter</span><span class="p">:</span>
        <span class="n">reads</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">reads_iter</span><span class="p">)</span>
        <span class="n">read_genes</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">infos</span> <span class="o">=</span> <span class="n">qname</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>

        <span class="n">barcode</span> <span class="o">=</span> <span class="n">infos</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">UMI</span> <span class="o">=</span> <span class="n">infos</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">barcode</span> <span class="o">!=</span> <span class="n">previous_barcode</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">previous_barcode</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="n">previous_barcode</span> <span class="o">=</span> <span class="n">barcode</span>
                <span class="k">continue</span>
            <span class="n">res2</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
                <span class="n">res2</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="n">gene</span><span class="p">])</span>
            <span class="n">res</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
                <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">previous_barcode</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
                <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">res2</span><span class="p">,</span> <span class="n">outfile</span><span class="p">)</span>
                <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

            <span class="n">previous_barcode</span> <span class="o">=</span> <span class="n">barcode</span>

        <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">reads</span><span class="p">:</span>
            <span class="c1">#Get Gene For A single Alignment</span>
            <span class="c1">#For each Alignment Of this Reads, we need to Get the gene Corresponed</span>
            <span class="c1">#If there are multiple Gene, return None</span>
            <span class="n">gene_info</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">rlen</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">rlen</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">read</span><span class="o">.</span><span class="n">is_unmapped</span><span class="p">:</span>
                <span class="c1">#There are multiple transcript For this Read</span>
                <span class="c1">#For each transcript, Get the Exon Coverage</span>
                <span class="c1">#If there are only One Gene, return the gene name if exon overlap ratio &gt;50%</span>
                <span class="c1">#If there are multiple Genes</span>

                <span class="n">tx_hits</span> <span class="o">=</span> <span class="n">align_to_transcriptome</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">tid</span><span class="p">,</span> <span class="n">read</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span> <span class="n">read</span><span class="o">.</span><span class="n">alen</span><span class="p">)</span>

                <span class="c1">#Filter the tx_hits using the Strand Information</span>
                <span class="c1">#If the Read is is_reversed, the strand of transcript should be 2</span>
                <span class="n">required_strand</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">is_reverse</span><span class="p">:</span>
                    <span class="n">required_strand</span> <span class="o">=</span> <span class="mi">2</span>

                <span class="c1"># tx_hits:</span>
                <span class="n">tx_hits</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span><span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">tx_hits</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">v</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="n">required_strand</span><span class="p">}</span>
                <span class="n">strand_info</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]:</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tx_hits</span><span class="o">.</span><span class="n">values</span><span class="p">()}</span>

                <span class="c1">#gene_info: initiate the informations for each gene in the tx_hits</span>
                <span class="n">gene_info</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]:{</span><span class="s1">&#39;overlap_len&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;dist_to_end&#39;</span><span class="p">:</span><span class="mi">10000</span><span class="p">}</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tx_hits</span><span class="o">.</span><span class="n">values</span><span class="p">()}</span>

                <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="n">tx_hits</span><span class="p">:</span>
                    <span class="n">gene</span> <span class="o">=</span> <span class="n">tx_hits</span><span class="p">[</span><span class="n">tx</span><span class="p">][</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span>
                    <span class="c1"># If the gene already have good alignment, skip</span>
                    <span class="k">if</span> <span class="n">gene_info</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s1">&#39;overlap_len&#39;</span><span class="p">]</span><span class="o">&gt;=</span> <span class="mf">0.9</span> <span class="o">*</span> <span class="n">rlen</span><span class="p">:</span>
                        <span class="k">continue</span>

                    <span class="n">alns</span> <span class="o">=</span> <span class="n">get_cigar_segments</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">cigar</span><span class="p">,</span> <span class="n">tx_hits</span><span class="p">[</span><span class="n">tx</span><span class="p">][</span><span class="s1">&#39;start&#39;</span><span class="p">])</span>
                    <span class="n">read_exon_overlap</span> <span class="o">=</span> <span class="mi">0</span>

                    <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="n">alns</span><span class="p">:</span>
                        <span class="n">read_exon</span> <span class="o">=</span> <span class="n">align2exon</span><span class="p">(</span><span class="n">aln</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">aln</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">tx_hits</span><span class="p">[</span><span class="n">tx</span><span class="p">][</span><span class="s1">&#39;strand&#39;</span><span class="p">],</span>
                                               <span class="n">tx_hits</span><span class="p">[</span><span class="n">tx</span><span class="p">][</span><span class="s1">&#39;exonIdx&#39;</span><span class="p">],</span> <span class="n">tx_hits</span><span class="p">[</span><span class="n">tx</span><span class="p">][</span><span class="s1">&#39;exonIdx&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">tx_hits</span><span class="p">[</span><span class="n">tx</span><span class="p">][</span><span class="s1">&#39;exons&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
                        <span class="n">read_exon_overlap</span> <span class="o">+=</span> <span class="n">read_exon</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>

                    <span class="n">gene_info</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s1">&#39;overlap_len&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">read_exon_overlap</span><span class="p">,</span> <span class="n">gene_info</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s1">&#39;overlap_len&#39;</span><span class="p">])</span>
                    <span class="n">gene_info</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="s1">&#39;dist_to_end&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">read_exon</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>

                <span class="n">gene_info</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span><span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">gene_info</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">v</span><span class="p">[</span><span class="s1">&#39;overlap_len&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mf">0.5</span><span class="o">*</span><span class="n">rlen</span><span class="p">}</span>

            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">gene_info</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">read_genes</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">v</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">read_genes</span><span class="p">))</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">mapgene</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">read_genes</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">gene_info</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">read_genes</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">v</span><span class="p">[</span><span class="s1">&#39;dist_to_end&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">400</span><span class="p">}</span>
            <span class="n">gene_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">gene_info</span><span class="o">.</span><span class="n">items</span><span class="p">()),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">val</span><span class="p">:</span> <span class="n">val</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;dist_to_end&#39;</span><span class="p">])]</span>
            <span class="n">gene_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="n">mapgene</span> <span class="o">=</span> <span class="n">gene_names</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">mapgene</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="k">if</span> <span class="n">mapgene</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
            <span class="n">res</span><span class="p">[</span><span class="n">mapgene</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">UMI</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">res</span><span class="p">[</span><span class="n">mapgene</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">UMI</span><span class="p">]</span>

    <span class="n">bam_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">res2</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
        <span class="n">res2</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="n">gene</span><span class="p">])</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
        <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">barcode</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">res2</span><span class="p">,</span> <span class="n">outfile</span><span class="p">)</span>
        <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span></div>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h3><a href="../../../index.html">Table Of Contents</a></h3>

<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">baseqDrops 0.01 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright BEISEQ Information Technology.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.7.4.
    </div>
  </body>
</html>